File: Java/(Hard)PermutationSequence.java
// Permutation Sequence

// The set [1, 2, 3, ..., n] contains a total of n! unique permutations.

// By listing and labeling all of the permutations in order, we get the following sequence for n = 3:
// "123"
// "132"
// "213"
// "231"
// "312"
// "321"
// Given n and k, return the kth permutation sequence.

// Example 1:
// Input: n = 3, k = 3
// Output: "213"

// Example 2:
// Input: n = 4, k = 9
// Output: "2314"

// Example 3:
// Input: n = 3, k = 1
// Output: "123"


//Solution from Copilot (correct)
//Big O time complexity O(n)
class Solution {
    public String getPermutation(int n, int k) {
        int[] factorial = new int[n];
        factorial[0] = 1;
        for (int i = 1; i < n; i++) {
            factorial[i] = factorial[i - 1] * i;
        }
        List<Integer> numbers = new ArrayList<>();
        for (int i = 1; i <= n; i++) {
            numbers.add(i);
        }
        k--;
        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= n; i++) {
            int index = k / factorial[n - i];
            sb.append(numbers.get(index));
            numbers.remove(index);
            k -= index * factorial[n - i];
        }
        return sb.toString();
    }
}



//Solution from ChatGPT (correct)
//Big O time complexity O(n)
class Solution {
    public String getPermutation(int n, int k) {
        List<Integer> nums = new ArrayList<>();
        int[] fact = new int[n + 1];
        StringBuilder sb = new StringBuilder();
        
        for (int i = 1; i <= n; i++) {
            nums.add(i);
        }
        
        fact[0] = 1;
        for (int i = 1; i <= n; i++) {
            fact[i] = i * fact[i - 1];
        }
        
        k--;
        
        for (int i = n - 1; i >= 0; i--) {
            int index = k / fact[i];
            sb.append(nums.get(index));
            nums.remove(index);
            k -= index * fact[i];
        }
        
        return sb.toString();
    }
}

File: JS/(Hard)PermutationSequence.js
// Permutation Sequence

// The set [1, 2, 3, ..., n] contains a total of n! unique permutations.

// By listing and labeling all of the permutations in order, we get the following sequence for n = 3:
// "123"
// "132"
// "213"
// "231"
// "312"
// "321"
// Given n and k, return the kth permutation sequence.

// Example 1:
// Input: n = 3, k = 3
// Output: "213"

// Example 2:
// Input: n = 4, k = 9
// Output: "2314"

// Example 3:
// Input: n = 3, k = 1
// Output: "123"

/**
 * @param {number} n
 * @param {number} k
 * @return {string}
 */


// Solution from ChatGPT (correct)
// Big O time complexity: O(n)
var getPermutation = function(n, k) {
    let nums = [];
    for (let i = 1; i <= n; i++) {
        nums.push(i);
    }
    let result = "";
    let factorial = 1;
    for (let i = 1; i <= n; i++) {
        factorial *= i;
    }
    k--;
    for (let i = 0; i < n; i++) {
        factorial /= n - i;
        let index = Math.floor(k / factorial);
        result += nums[index];
        nums.splice(index, 1);
        k -= index * factorial;
    }
    return result;
};



// Solution from Copilot (correct)
// Big O time complexity: O(n)
var getPermutation = function(n, k) {
    let nums = [];
    let factorial = 1;
    
    for (let i = 1; i <= n; i++) {
        nums.push(i);
        factorial *= i;
    }
    
    k--;
    let result = "";
    
    for (let i = 0; i < n; i++) {
        factorial /= n - i;
        let index = Math.floor(k / factorial);
        result += nums[index];
        nums.splice(index, 1);
        k -= index * factorial;
    }
    
    return result;
};